packages/nx/src/native/tui/app.rs

```diff
@@ -774,17 +939,442 @@ impl App {
+                // Unpin the task if it's already pinned
+                self.pane_tasks[pane_idx] = None;
+
+                // If this was previously pane 2 and its now unpinned and pane 1 is still set, set the pane arrangement to single
+                if pane_idx == 1 && self.pane_tasks[0].is_some() {
+                    self.layout_manager
+                        .set_pane_arrangement(PaneArrangement::Single);
+                }
+
+                // Adjust focused pane if necessary
+                if !self.has_visible_panes() {
+                    self.update_focus(Focus::TaskList);
+                    // When all panes are cleared, use position-based selection
+                    self.set_spacebar_mode(false, Some(SelectionMode::TrackByPosition));
+                    // If no visible panes are left, set the pane arrangement to none
+                    self.layout_manager
+                        .set_pane_arrangement(PaneArrangement::None);
+                }
+
+                self.dispatch_action(Action::UnpinTask(task_name.clone(), pane_idx));
+            } else {
+                // Pin the task to the specified pane
+                self.pane_tasks[pane_idx] = Some(task_name.clone());
+                self.update_focus(Focus::TaskList);
+
+                // Exit spacebar mode when pinning
+                // When pinning a task, use name-based selection
+                self.set_spacebar_mode(false, Some(SelectionMode::TrackByName));
+
+                if self.layout_manager.get_pane_arrangement() == PaneArrangement::None {
+                    if pane_idx == 0 && self.pane_tasks[1].is_none() {
+                        self.layout_manager
+                            .set_pane_arrangement(PaneArrangement::Single);
+                    } else if pane_idx == 1 || pane_idx == 0 && self.pane_tasks[1].is_some() {
+                        self.layout_manager
+                            .set_pane_arrangement(PaneArrangement::Double);
+                    }
+                }
+
+                self.dispatch_action(Action::PinTask(task_name.clone(), pane_idx));
+            }
+        }
+
+        // Always re-evaluate the optimal size of the terminal pane(s)
+        self.recalculate_layout_areas();
+        // Ensure the pty instances get resized appropriately (no debounce as this is based on an imperative user action)
+        let _ = self.handle_pty_resize();
+    }
+
+    /// Forward key events to the currently focused pane, if any.
+    fn handle_key_event(&mut self, key: KeyEvent) -> io::Result<()> {
+        if let Focus::MultipleOutput(pane_idx) = self.focus {
+            let terminal_pane_data = &mut self.terminal_pane_data[pane_idx];
+            terminal_pane_data.handle_key_event(key)
+        } else {
+            Ok(())
+        }
+    }
+
+    /// Returns true if the currently focused pane is in interactive mode.
+    fn is_interactive_mode(&self) -> bool {
+        match self.focus {
```

Comment:
future: I think interactive mode belongs on the app even though the terminal pane is receiving the inputs. Interactive mode means at the app level, almost all key events have to get sent into the pane. So rather than checking the pane to see if it's interactive, I think it's easier if we track interactive mode in the app.rs